### ISSUE TITLE
Current span in Tracer lost after WebClient call

### ISSUE THREAD

**Issue by ITman1** [micrometer-metrics/micrometer#3785]
**Describe the bug**
- I do call `Tracer.currentSpan()` before and after blocking `WebClient` call. 
- I would expect the same span to be present still in there after the call. Instead of that I am getting `null`.
- I do use `Hooks.enableAutomaticContextPropagation()`, without this hook propagation does not work at all, even if I used `.contextCapture()` before `.block()` call.

**My test**
![image](https://user-images.githubusercontent.com/1372054/233663817-04b2a974-0c3b-480c-ae4c-26069d8952f5.png)

**Environment**
 - Spring Boot 3.0.6

**To Reproduce**
Here is demo repository: https://github.com/ITman1/webclient-enableAutomaticContextPropagation-span-lost

**Comment by xerx593** [comment]
### "Fix proposal":
```java
@Test
    void spanIdToParentId() {
        TraceContext contextBeforeCall = tracer.currentSpan().context();
        Mono<String> responseTraceIds = blockingWebClientCall(); // stay reactive!
        TraceContext contextAfterCall = Optional.ofNullable(tracer.currentSpan()).map(Span::context).orElse(null);

        Assertions.assertEquals(traceId + "-" + traceId, responseTraceIds.block(), "response should contain trace ID from Tracer and from MDC");
        Assertions.assertEquals(contextBeforeCall, contextAfterCall, "current context before and after blocking call should be the same");
    }

    private Mono<String> blockingWebClientCall() {
       
        return webClientBuilder
                .build()
                .get()
                .uri("http://localhost:" + port + "/trace-id")
                .header("hcitrace", "some-hci-trace")
                .retrieve()
                .bodyToMono(String.class)
                .publishOn(Schedulers.immediate()); // ?
    }
```

**Comment by ITman1** [comment]
@xerx593 It is not so big surprise to me that tests will start to pass if you postpone actual subscription to response from `WebClient`, it is like doing nothing when `WebClient` call is actually deffered by the nature. Just try to put `tracer.currentSpan()` after `StepVerifier` and you will get `null` again.

This behaviour (context equality before/after blocking call) used to work in Sleuth implementation, actually it worked also after migration to Micrometer implementation in our company tests, but stopped to work after enabling `enableAutomaticContextPropagation` hook because we did not want to wrap each operator with `ContextSnapshot.setAllThreadLocalsFrom`. I would like to demonstrate also that this worked when not using `enableAutomaticContextPropagation` hook, but last time when I attempted to do that I failed, I was not able to reproduce it from our company tests into empty demo project, but I will try again in the next week(s).

About bad design/mis-use I duly agree with you, but this is only example from real world where we unfornatelly have to mix blocking and non-blocking approaches, eg. because of missing non-blocking I/O drivers to certain services.
